pybind11 v3 breaks unit test

It looks like [pybind11 v3.0](https://github.com/pybind/pybind11/releases/tag/v3.0.0) just released. As with any major version bump, there are breaking changes...

Our unit test is setup up to detect changes like this, and it has:

<details><summary>Logs about the test failure</summary>
<p>

```
  =================================== FAILURES ===================================
  ______________________ test_pybind11_overloaded_function _______________________
  
  apigen_make_app = <function apigen_make_app.<locals>.make at 0x7fbedab929e0>
  snapshot = <pytest_snapshot.plugin.Snapshot object at 0x7fbedb1405e0>
  
      def test_pybind11_overloaded_function(apigen_make_app, snapshot):
          testmod = "sphinx_immaterial_pybind11_issue_134"
          app = apigen_make_app(
              confoverrides=dict(
                  python_apigen_modules={
                      testmod: "api/",
                  },
              ),
          )
          print(app._status.getvalue())
          print(app._warning.getvalue())
          assert not app._warning.getvalue()
      
          data = _get_api_data(app.env)
      
          def get_entity_info(key):
              entity = data.entities[key]
              return json.dumps(
                  {
                      k: getattr(entity, k)
                      for k in ["signatures", "primary_entity", "siblings"]
                  },
                  indent=2,
              )
      
          for name in [
              "Example.foo(int)",
              "Example.foo(bool)",
              "Example.bar(int)",
              "Example.bar(bool)",
          ]:
  >           snapshot.assert_match(
                  get_entity_info(f"{testmod}.{name}"),
                  name.replace("(", "_").replace(")", "") + ".json",
              )
  E           AssertionError: value does not match the expected value in snapshot tests/snapshots/python_apigen_test/test_pybind11_overloaded_function/Example.foo_int.json
  E             (run pytest with --snapshot-update to update snapshots)
  E           assert '{\n  "signatures": [\n    "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: typing.SupportsInt) -> int"\n  ],\n  "primary_entity": true,\n  "siblings": {\n    "sphinx_immaterial_pybind11_issue_134.Example.bar(int)": true\n  }\n}' == '{\n  "signatures": [\n    "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: int) -> int"\n  ],\n  "primary_entity": true,\n  "siblings": {\n    "sphinx_immaterial_pybind11_issue_134.Example.bar(int)": true\n  }\n}'
  E             
  E               {
  E                 "signatures": [
  E             -     "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: int) -> int"
  E             +     "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: typing.SupportsInt) -> int"
  E             ?                                                                 +++ ++++++++++++
  E                 ],
  E                 "primary_entity": true,
  E                 "siblings": {
  E                   "sphinx_immaterial_pybind11_issue_134.Example.bar(int)": true
  E                 }
  E               }
  
  tests/python_apigen_test.py:222: AssertionError
  =============================== warnings summary ===============================
  tests/python_apigen_test.py::test_type_params
    /home/runner/work/sphinx-immaterial/sphinx-immaterial/.nox/tests-py3-10-sphinx8/lib/python3.10/site-packages/sphinx/util/docfields.py:107: DeprecationWarning: nodes.Text: initialization argument "rawsource" is ignored and will be removed in Docutils 2.0.
      refnode += contnode or innernode(target, target)  # type: ignore[call-arg]
``` 

</p>
</details> 

In a perfect world, we'd support pybind11 v2 and v3.

I think we have some time on this one. I doubt the pybind11 users are rushing to upgrade their bindings to use v3, but I've been wrong before.

--- a/tests/snapshots/python_apigen_test/test_pybind11_overloaded_function/Example.bar_int.json
+++ b/tests/snapshots/python_apigen_test/test_pybind11_overloaded_function/Example.bar_int.json
@@ -1,6 +1,6 @@
 {
   "signatures": [
-    "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: int) -> int"
+    "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: typing.SupportsInt) -> int"
   ],
   "primary_entity": false,
   "siblings": null

--- a/tests/snapshots/python_apigen_test/test_pybind11_overloaded_function/Example.foo_int.json
+++ b/tests/snapshots/python_apigen_test/test_pybind11_overloaded_function/Example.foo_int.json
@@ -1,6 +1,6 @@
 {
   "signatures": [
-    "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: int) -> int"
+    "(self: sphinx_immaterial_pybind11_issue_134.Example, arg0: typing.SupportsInt) -> int"
   ],
   "primary_entity": true,
   "siblings": {

